src/main/java/org/terasology/math/geom/AABB.java

```diff
@@ -0,0 +1,477 @@
+                minDistance = distance;
+                closestNormal = n;
+            }
+        }
+
+        return closestNormal;
+    }
+
+    /**
+     * Returns the center point of one of the six planes for the given normal.
+     *
+     * @param normal The normal
+     * @return The center point
+     */
+    public Vector3f centerPointForNormal(Vector3f normal) {
+        if (normal.x == 1 && normal.y == 0 && normal.z == 0) {
+            return new Vector3f(max.x, getCenter().y, getCenter().z);
+        }
+        if (normal.x == -1 && normal.y == 0 && normal.z == 0) {
+            return new Vector3f(min.x, getCenter().y, getCenter().z);
+        }
+        if (normal.x == 0 && normal.y == 0 && normal.z == 1) {
+            return new Vector3f(getCenter().x, getCenter().y, max.z);
+        }
+        if (normal.x == 0 && normal.y == 0 && normal.z == -1) {
+            return new Vector3f(getCenter().x, getCenter().y, min.z);
+        }
+        if (normal.x == 0 && normal.y == 1 && normal.z == 0) {
+            return new Vector3f(getCenter().x, max.y, getCenter().z);
+        }
+        if (normal.x == 0 && normal.y == -1 && normal.z == 0) {
+            return new Vector3f(getCenter().x, min.y, getCenter().z);
+        }
+
+        return new Vector3f();
+    }
+
+    public float minX() {
+        return min.x;
+    }
+
+    public float minY() {
+        return min.y;
+    }
+
+    public float minZ() {
+        return min.z;
+    }
+
+    public float maxX() {
+        return max.x;
+    }
+
+    public float maxY() {
+        return max.y;
+    }
+
+    public float maxZ() {
+        return max.z;
+    }
+
+    /**
+     * Returns the vertices of this AABB.
+     *
+     * @return The vertices
+     */
+    public Vector3f[] getVertices() {
+        if (vertices == null) {
+            vertices = new Vector3f[8];
+
+            // Front
+            vertices[0] = new Vector3f(min.x, min.y, max.z);
+            vertices[1] = new Vector3f(max.x, min.y, max.z);
+            vertices[2] = new Vector3f(max.x, max.y, max.z);
+            vertices[3] = new Vector3f(min.x, max.y, max.z);
+            // Back
+            vertices[4] = new Vector3f(min.x, min.y, min.z);
+            vertices[5] = new Vector3f(max.x, min.y, min.z);
+            vertices[6] = new Vector3f(max.x, max.y, min.z);
+            vertices[7] = new Vector3f(min.x, max.y, min.z);
+        }
+
+        return vertices;
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (this == obj) {
+            return true;
+        }
+        if (obj instanceof AABB) {
+            AABB other = (AABB) obj;
+            return Objects.equal(min, other.min) && Objects.equal(max, other.max);
+        }
+        return false;
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hashCode(min, max);
+    }
+
+    public boolean intersectRectangle(Vector3f from, Vector3f direction) {
+        Vector3f dirFrac = new Vector3f(
```

Comment:
if at the end only the relation between tmin and tmax counts it might not be necessary to normalize the vectors, just take their absolute value so the signs will be the same